### ISSUE TITLE
Support for testing libraries

### ISSUE THREAD

**Issue by Dru89** [raycast/extensions#1803]
Raycast version: 1.35.1

**Describe the feature and the current behavior/state.**
It would be great to have some way to write some kind of unit/integration tests for my extensions with [`jest`](https://jestjs.io) and [`@testing-library/react`](https://testing-library.com/docs/react-testing-library/intro) or something similar.

I was about to refactor an extension, but I wanted to make sure the behavior would remain the same. I started down the road of setting up some testing and made a *little* progress that you can see from the diff of my test branch: https://github.com/raycast/extensions/compare/main...Dru89:unit-test-example

The problem comes down to the test file. Consider something like:
```ts
import { render } from "@testing-library/react";
import { Detail } from "@raycast/api";

it("can render a Detail", () => {
  const { debug } = render(<Detail markdown="**some markdown**" />);
  debug(); // should print the DOM to the screen
  expect(true).toBe(true);
});
```

This relatively simple test fails since `import { Detail } from "@raycast/api"` doesn't actually "point" to anything other than the typescript definition in the packaged node module.

I suspect in order to make something work here with the existing setup, we'd need something like a `ray test` that did all of the bootstrapping of the test environment for us.

I also am unsure about whether or not `@testing-library/react` would work out of the box after this, or if we'd need something else. [Testing Library](https://testing-library.com) expects there to be a (virtual) browser DOM underneath each component and I'm not clear if the various components like `<Detail />` are doing something like that or are instead rendering into "native" objects. There's also [React Native Testing Library](https://callstack.github.io/react-native-testing-library/) that might help here, but again I'm not sure.

If this is something that could be community driven, I think that would be totally ok! It's just that I don't think `@raycast/api` currently exposes the hooks that we'd need to get that far so I wanted to raise this as a dev tools feature request.

**Comment by mathieudutour** [comment]
Good timing, I'm working on something like this :)

You won't be able to use `@testing-library/react` because it's for react-dom, same for `@testing-library/react-native`.

Here is what it's probably going to look like:
```js
import { List, ActionPanel, Action, Detail } from "@raycast/api";
import { create } from "@raycast/apitest-renderer";

describe("Push", () => {
  it.only("should push a new view at the top level", async () => {
    const renderer = await create(
      <List>
        <List.Item
          title="first item"
          actions={
            <ActionPanel>
              <Action.Push title="push" target={<Detail markdown="detail" />} />
            </ActionPanel>
          }
        />
      </List>
    );

    expect(renderer.views.length).toBe(1);
    expect(renderer.views[0].type).toBe(List);

    const action = renderer.views[0].findByType(Action);

    await renderer.fireEvent(action, "action");

    expect(renderer.views.length).toBe(2);
    expect(renderer.views[0].type).toBe(List);
    expect(renderer.views[1].type).toBe(Detail);
  });
});
```

**Comment by Dru89** [comment]
That sounds great, thanks!

Sorry if this was already covered by some other issue. I'd tried to search through the issues and Slack, but didn't find anything related yet. 😅

**Comment by Rawnly** [comment]
Hello! Any progress on this? I'd like to test some utility functions but I currently have some issues with the `@raycast/api` package

Currently using `vitest` with a minimal setup.

![CleanShot 2022-09-19 at 17 59 50](https://user-images.githubusercontent.com/16429579/191061491-3549b61d-85b8-4500-8b4e-9cb90b5b9ea9.png)

**Comment by jose-elias-alvarez** [comment]
I stumbled upon this issue after a lot of trial and error, and I've at least managed to get around the `import { x } from "@raycast/api"` issue with a Jest [manual mock](https://jestjs.io/docs/manual-mocks):

1. Create a `__mocks__/@raycast` directory at the root of the project (adjacent to `node_modules`)
2. Create an `api.js` file in the directory:

```javascript
// I specifically used a plain .js file to avoid having to disable `esModuleInterop`, 
// but it's not strictly necessary
// module.exports syntax *does* appear to be necessary, at least with a default ts-jest setup
module.exports = {
  // mock whatever API exports you need access to
  getPreferenceValues: jest.fn(),
  getSelectedFinderItems: jest.fn(),
  Toast: {
    Style: {},
  },
  showToast: jest.fn(),
};
```

3. Since this is a module, it's automatically mocked in tests. If you need to do something with the mock, e.g. mock a return / resolved value, you can do so in the usual way:

```typescript
import { getSelectedFinderItems } from "@raycast/api";

(getSelectedFinderItems as jest.Mock).mockResolvedValue([
  {
    path: "/my/mock/file/path",
  },
]);
```

I imagine any sort of rendering logic will significantly complicate things (the plugin I'm working on is quite simple and doesn't use React) but since the approach above unblocked me on unit testing, I thought I'd share.
